### PR TITLE
Implement Api fns for arbitrary subresources and approval subresource for Api<CertificateSigningRequest>

### PR DESCRIPTION
**Pull request by ChinYing-Li**
## Motivation
This PR tackles two tasks in #127.

## Solution
Implement Api functions for arbitrary subresource and Api<CertificateSigningRequest> approve/deny.
One test case of Api<CertificateSigningRequest>  is added.
Should I try to come up with test cases for those `Api` subresource-related functions?
IMO,  testing the logic of `Request.*_subresource` should be sufficient.
Any suggestion is appreciated!

### CODE REVIEW

File: kube-client/src/api/util.rs
Comment (clux) [inline kube-client/src/api/util.rs:76]: hm, while i do think these look like mostly sensible defaults. it's probably up to the controller doing this to attribute meaning to these fields. it's also slightly outside the purview of the api to be setting defaults here, it's something more normally done in the runtime crate.

personally, i think we should probably only have one method exposed from `Api<CertificateSigningRequest>`, and we should probably take a `CertificateSigningRequestStatus` directly. Maybe `::sign_respond`.

File: kube-client/src/api/util.rs
Comment (clux) [inline kube-client/src/api/util.rs:81]: we are also taking `PatchParams` in this fn, but always using `Patch::strategic`, so this is not ideal.

File: kube-client/src/api/subresource.rs
Comment (clux) [inline kube-client/src/api/subresource.rs:63]: I like these :+1:

**Review by clux — COMMENTED** [review]
hey thanks again for all these prs!

i think the general setup here is ok, but have left a few comments on the main `Api<CSR>` abstraction. I feel it is trying to make a few too many assumptions for what should be in `kube-client`, and it's probably better to keep this simple and just wrap the newly added `patch_subresource` method.

(however, there may be an argument here for replicating what [kubectl certificate approve](https://github.com/kubernetes/kubectl/blob/d7da6ad9f193e3afdc754527c0a5ac9390c053fb/pkg/cmd/certificates/certificates.go#L156-L164) as a small `certificate` helper in `kube-runtime` - but that is not necessary for this pr - and not even sure we need to do it).

File: kube-client/src/api/util.rs
Comment (ChinYing-Li) [reply kube-client/src/api/util.rs:81]: I guess we should use Patch::Merge then. Please let me know if I am wrong!

File: kube-client/src/api/util.rs
Comment (ChinYing-Li) [reply kube-client/src/api/util.rs:76]: I think it would be more straight-forward to just  call the fn `patch_approval`, WDYT?

Comment (clux) [reply kube-client/src/api/util.rs]: Ah, yeah, that sounds good :+1:

File: kube-client/src/api/util.rs
Comment (clux) [reply kube-client/src/api/util.rs]: The problem is that we let users set their own `PatchParams`, but unlike `Api::patch`, they don't actually supply a `Patch` themselves, so it's unlike anything we have and doesn't match conventions.

Options:

1. name the function `set_approval`, and just not expose the `PatchParams`
2. take a `Patch<T>` in `patch_approval` instead of a specific struct (but have a doc example use a `CertificateSigningRequest`

1 is arguably more ergonomic, but it makes choices that we probably should not be making in `kube-client`.
Otoh, 2 is just a dumb wrapper for `Api::patch_subresource("approval", ...)`, but this plays into goals for generic subresource handling nicer down the line.

Personally, I think we should follow the dumb approach here because it matches how we do `get_scale` / `patch_scale` / `get_status` / `patch_status`. If we do the dumb options, then we can also do:

- `Api::get_approval`
- `Api::replace_approval`

AFAIKT we could lock down the types for these 2 (like `Scale`) because it looks they return a full `CertificateSigningRequest` (so the subresource is kind of the same struct as itself, the part it cares about is just a subset of it..).

---

side-notes on generic subresource handling: background info https://github.com/kube-rs/k8s-pb/issues/4 and discussion in https://github.com/Arnavion/k8s-openapi/pull/98#issuecomment-853439773 (although this has been put on hold for a bit to try to tackle it direct in k8s-pb)

If the Approval subresource is not its own struct but reusing the root one, then the subresource behaves like status and not scale. This makes one of the original subresource plans kind of awkward: `Api<CertificateSigningRequest>::subresource<CertificateSigningRequest>::patch()` is rather ugly, so it's perhaps better to stick with the longer names like `Api<CertificateSigningRequest>::patch_approval()` long term - and find a way to generate this in the future)

Comment (ChinYing-Li) [reply kube-client/src/api/util.rs]: Got it, thank you for the explanation!

Comment (clux) [reply kube-client/src/api/util.rs]: Thanks a lot - and for your patience during the new year :-)
Everything in the PR is great and happy to merge it as is.

Just have one question left. Do you think there's a benefit to adding `Api::get_approval` while we are here?
AFAIKT it would just do the same thing as `Api::get` for the CSR (because it's the same struct), but technically [kubernetes supports the GET verb on the approval subresource](https://github.com/kube-rs/k8s-pb/blob/d60271637aac5bb854d529f99d5e6a23154acd06/k8s-pb-codegen/openapi/transformed.json#L804-L809).

**Review by clux — APPROVED** [review]

File: kube-client/src/api/util.rs
Comment (ChinYing-Li) [reply kube-client/src/api/util.rs]: Thank you for reviewing during the holidays! I'd be happy to add `Api<CSR>::get_approval` if that's useful, but isn't it just another wrapper of `get`:
```rust
pub async fn get_approval(
        &self,
        name: &str) -> Result<CertificateSigningRequest> {
      self.get(name)
}
```

Comment (clux) [reply kube-client/src/api/util.rs]: I think if we do it, it should be a wrapper for your new `get_subresource("approval", name)` because then it's up to the apiserver whether or not it's actually handled differently (rather than us making an alias for consistency).

But I don't think we need to do it if this actually returns the same thing as `get(name)` which i suspect it does. If you are able to verify, then that would be good to know (and maybe good to leave a comment about for the next person).

Comment (ChinYing-Li) [reply kube-client/src/api/util.rs]: Interestingly, while both `get(csr_name)` and `get_subresource("approval", csr_name)` returns a CertificateSigningRequest, the returned CSR differs in two places: 
1. `resource_version`
2. `status.certificate`: it's `none` for `get(name)` but not none for `get_subresource("approval", csr_name)`.

Comment (clux) [reply kube-client/src/api/util.rs]: Ah, interesting! Then we should definitely also export an alias `get_approval` as an alias for `get_subresource("approval", ...)`

Comment (ChinYing-Li) [reply kube-client/src/api/util.rs]: Just added that!

File: kube-client/src/lib.rs
Comment (clux) [inline kube-client/src/lib.rs:509]: i think should set an explicit fake certificate, and use `csr.get_approval` to test the new method (since it will include it in the status)

File: kube-client/src/api/util.rs
Comment (clux) [reply kube-client/src/api/util.rs]: Looks great. Left a minor follow-up comment on the test (which is still using `.get`).

**Comment by clux** [comment]
This looks great now. Thanks for doing all the minor tweaks here to get it into a great state! I'll merge this in tomorrow if you are happy with it :-)

**Comment by ChinYing-Li** [comment]
I think this is ready for merge. Thanks for helping!

**Pull request merged**
